### PR TITLE
Blockly Factory: Shadow Block Validation in Workspace Factory

### PR DESCRIPTION
**Pull request by evd2014**
Increased shadow block validation in workspace factory: 

- When converting a block to a shadow block, all of its children are also converted to shadow blocks (to help users avoid invalid state)

- Provide warning message when a block with a variable field is made a shadow block.

- Provide warning message when a normal block is nested inside a shadow block. 

shadow block warnings

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/563)

### CODE REVIEW

**Comment by evd2014** [comment]
@picklesrus @vicng @quachtina96

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563#-:-KPPObcU_4ru3EN1QbWQ:bvyui8)*

**Comment by picklesrus** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_init.js, line 342 \[r1\]](https://reviewable.io:443/reviews/google/blockly/563#-KPPqwUuciixz6wAp3Pe:-KPPqwUuciixz6wAp3Pf:b-9cwyxw) ([raw file](https://github.com/google/blockly/blob/7dc0dfd10d57344aac441233f24ecaf175c7844e/demos/blocklyfactory/workspacefactory/wfactory_init.js#L342)):*
> ```JavaScript
>             ((controller.isUserGenShadowBlock(selected.id) &&
>             !selected.getSurroundParent()) ||
>             (!controller.isUserGenShadowBlock(selected.id) &&
> ```

This conditional is kind of long and hard to parse. Maybe it throw it into a private method with a descriptive name so it reads a bit better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563)*

**Comment by evd2014** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_init.js, line 342 \[r1\]](https://reviewable.io:443/reviews/google/blockly/563#-KPPqwUuciixz6wAp3Pe:-KPQ4W9aafdMi1qzEWgF:b2fjy0r) ([raw file](https://github.com/google/blockly/blob/7dc0dfd10d57344aac441233f24ecaf175c7844e/demos/blocklyfactory/workspacefactory/wfactory_init.js#L342)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> This conditional is kind of long and hard to parse. Maybe it throw it into a private method with a descriptive name so it reads a bit better?
</details>
Moved it into an inline function so that it reads more easily (left it inlined because it's specific to the function).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563)*

**Comment by quachtina96** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_init.js, line 380 \[r4\]](https://reviewable.io:443/reviews/google/blockly/563#-KPQAW016HnyYMlqa5N8:-KPQAW016HnyYMlqa5N9:b-k55mbn) ([raw file](https://github.com/google/blockly/blob/71a2cdf60e53312b44c3eff79ad87a20e7c5492f/demos/blocklyfactory/workspacefactory/wfactory_init.js#L380)):*
> ```JavaScript
>   // parent, or (2) a normal block is inside of a shadow block.
>   var isInvalidBlockPlacement = function(block) {
>     return ((controller.isUserGenShadowBlock(block.id) &&
> ```

multiple if statements that check if its invalid. e.g. having an if statement for each rule

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563)*

**Comment by picklesrus** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563#-:-KPTpIHdSXKNHalyY-3-:bnfp4nl)*

**Comment by quachtina96** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563#-:-KPUQFclC12i63zptYkb:bnfp4nl)*

**Comment by vicng** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563#-:-KPUfzg-IWOsbt5QWFmH:bnfp4nl)*

**Comment by evd2014** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/workspacefactory/wfactory_init.js, line 380 \[r4\]](https://reviewable.io:443/reviews/google/blockly/563#-KPQAW016HnyYMlqa5N8:-KPUhPasGQ8Z7dRK2Ayx:b-mkg7u9) ([raw file](https://github.com/google/blockly/blob/71a2cdf60e53312b44c3eff79ad87a20e7c5492f/demos/blocklyfactory/workspacefactory/wfactory_init.js#L380)):*
<details><summary><i>Previously, quachtina96 (Tina Quach) wrote…</i></summary>
> multiple if statements that check if its invalid. e.g. having an if statement for each rule
</details>
We talked about this and not going to change this one!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/563)*
